File: modules/agent/src/serializers/notification.py
"""Serializers for notification-related models."""

from typing import Optional

from repositories.notification_repository import (
    get_entity_project_id_sync,
    get_lead_type_sync,
)

ENTITY_URL_MAP = {
    "Lead": "/leads",
    "Deal": "/deals",
    "Task": "/tasks",
    "Individual": "/accounts/individuals",
    "Organization": "/accounts/organizations",
}


def _get_entity_url(entity_type: str, entity_id: int, comment_id: Optional[int] = None) -> str:
    """Get URL for an entity, optionally with comment anchor."""
    if entity_type == "Lead":
        lead_type = get_lead_type_sync(entity_id)
        url = f"/leads/{lead_type}/{entity_id}"
    else:
        base_path = ENTITY_URL_MAP.get(entity_type, f"/{entity_type.lower()}s")
        url = f"{base_path}/{entity_id}"

    if not comment_id:
        return url
    return f"{url}#comment-{comment_id}"


def _get_entity_display_name(comment) -> str:
    """Get display name for the entity a comment belongs to."""
    return f"{comment.commentable_type} #{comment.commentable_id}"


def notification_to_response(notification) -> dict:
    """Convert notification model to dictionary with nested comment/entity info."""
    comment = notification.comment
    creator = comment.creator if comment else None

    created_by_name = None
    if creator:
        first = creator.first_name or ""
        last = creator.last_name or ""
        created_by_name = f"{first} {last}".strip() or creator.email

    content_preview = ""
    if comment and comment.content:
        content_preview = comment.content[:100] + ("..." if len(comment.content) > 100 else "")

    return {
        "id": notification.id,
        "type": notification.notification_type,
        "is_read": notification.is_read,
        "created_at": notification.created_at.isoformat() if notification.created_at else None,
        "comment": {
            "id": comment.id if comment else None,
            "content": content_preview,
            "created_by_name": created_by_name,
            "created_at": comment.created_at.isoformat() if comment and comment.created_at else None,
        } if comment else None,
        "entity": {
            "type": comment.commentable_type if comment else None,
            "id": comment.commentable_id if comment else None,
            "display_name": _get_entity_display_name(comment) if comment else None,
            "url": _get_entity_url(comment.commentable_type, comment.commentable_id, comment.id) if comment else None,
            "project_id": get_entity_project_id_sync(comment.commentable_type, comment.commentable_id) if comment else None,
        } if comment else None,
    }

File: modules/agent/internal/serializers/notification.go
package serializers

import (
	"fmt"
	"strings"
	"time"

	"agent/internal/models"
)

// EntityURLResolver provides methods to resolve entity URLs and project IDs
type EntityURLResolver interface {
	GetLeadType(entityID int64) string
	GetEntityProjectID(entityType string, entityID int64) *int64
}

// NotificationResponse represents a single notification
type NotificationResponse struct {
	ID        int64                    `json:"id"`
	Type      string                   `json:"type"`
	IsRead    bool                     `json:"is_read"`
	CreatedAt *string                  `json:"created_at"`
	Comment   *NotificationComment     `json:"comment"`
	Entity    *NotificationEntity      `json:"entity"`
}

// NotificationComment represents comment info in a notification
type NotificationComment struct {
	ID            *int64  `json:"id"`
	Content       string  `json:"content"`
	CreatedByName *string `json:"created_by_name"`
	CreatedAt     *string `json:"created_at"`
}

// NotificationEntity represents entity info in a notification
type NotificationEntity struct {
	Type        *string `json:"type"`
	ID          *int64  `json:"id"`
	DisplayName *string `json:"display_name"`
	URL         *string `json:"url"`
	ProjectID   *int64  `json:"project_id"`
}

// NotificationsResponse represents the list response
type NotificationsResponse struct {
	Notifications []NotificationResponse `json:"notifications"`
	UnreadCount   int64                  `json:"unread_count"`
}

var entityURLMap = map[string]string{
	"Lead":         "/leads",
	"Deal":         "/deals",
	"Task":         "/tasks",
	"Individual":   "/accounts/individuals",
	"Organization": "/accounts/organizations",
}

// NotificationToResponse converts a notification model to response
func NotificationToResponse(n *models.CommentNotification, resolver EntityURLResolver) NotificationResponse {
	resp := NotificationResponse{
		ID:     n.ID,
		Type:   n.NotificationType,
		IsRead: n.IsRead,
	}

	if !n.CreatedAt.IsZero() {
		t := n.CreatedAt.Format(time.RFC3339)
		resp.CreatedAt = &t
	}

	if n.Comment == nil {
		return resp
	}

	resp.Comment = buildNotificationComment(n.Comment)
	resp.Entity = buildNotificationEntity(n.Comment, resolver)

	return resp
}

func buildNotificationComment(c *models.Comment) *NotificationComment {
	comment := &NotificationComment{
		ID:      &c.ID,
		Content: truncateContent(c.Content, 100),
	}

	if !c.CreatedAt.IsZero() {
		t := c.CreatedAt.Format(time.RFC3339)
		comment.CreatedAt = &t
	}

	if c.Creator == nil {
		return comment
	}

	name := buildCreatorName(c.Creator)
	comment.CreatedByName = name

	return comment
}

func buildCreatorName(creator *models.User) *string {
	first := derefStr(creator.FirstName)
	last := derefStr(creator.LastName)
	name := strings.TrimSpace(first + " " + last)

	if name == "" {
		return &creator.Email
	}
	return &name
}

func buildNotificationEntity(c *models.Comment, resolver EntityURLResolver) *NotificationEntity {
	displayName := fmt.Sprintf("%s #%d", c.CommentableType, c.CommentableID)
	url := getEntityURL(c.CommentableType, c.CommentableID, &c.ID, resolver)
	projectID := resolver.GetEntityProjectID(c.CommentableType, c.CommentableID)

	return &NotificationEntity{
		Type:        &c.CommentableType,
		ID:          &c.CommentableID,
		DisplayName: &displayName,
		URL:         url,
		ProjectID:   projectID,
	}
}

func getEntityURL(entityType string, entityID int64, commentID *int64, resolver EntityURLResolver) *string {
	url := buildBaseEntityURL(entityType, entityID, resolver)
	url = appendCommentAnchor(url, commentID)
	return &url
}

func buildBaseEntityURL(entityType string, entityID int64, resolver EntityURLResolver) string {
	if entityType == "Lead" {
		leadType := resolver.GetLeadType(entityID)
		return fmt.Sprintf("/leads/%s/%d", leadType, entityID)
	}
	basePath := getEntityBasePath(entityType)
	return fmt.Sprintf("%s/%d", basePath, entityID)
}

func getEntityBasePath(entityType string) string {
	basePath, ok := entityURLMap[entityType]
	if ok {
		return basePath
	}
	return "/" + strings.ToLower(entityType) + "s"
}

func appendCommentAnchor(url string, commentID *int64) string {
	if commentID == nil {
		return url
	}
	return fmt.Sprintf("%s#comment-%d", url, *commentID)
}

func truncateContent(content string, maxLen int) string {
	if len(content) <= maxLen {
		return content
	}
	return content[:maxLen] + "..."
}

func derefStr(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}
